File: R/lgb_plus_A.R
#' LGB^A+: Hybrid Tree + Linear Boosting
#'
#' A gradient boosting variant that alternates between:
#' 1. Fitting a block of trees on residuals
#' 2. Fitting a single linear term (greedy feature selection) on residuals
#'
#' This captures both nonlinear patterns (via trees) and linear structure
#' (via repeated linear updates), making it robust to unknown DGPs.
#'
#' @author Philippe Goulet Coulombe (UQAM)
#' @license MIT

# Check and load required packages
.check_lightgbm <- function() {
  if (!requireNamespace("lightgbm", quietly = TRUE)) {
    stop("Package 'lightgbm' is required. Install with: install.packages('lightgbm')")
  }
}

#' Fit an LGB^A+ model
#'
#' @param X Matrix or data.frame of features (n_samples x n_features)
#' @param y Numeric vector of target values
#' @param n_cycles Number of alternating cycles (default: 25)
#' @param trees_per_cycle Trees to fit per cycle (default: 10)
#' @param lr_tree Learning rate for trees (default: 0.02)
#' @param lr_linear Learning rate for linear updates (default: 0.1)
#' @param num_leaves Max leaves per tree (default: 15)
#' @param min_data Minimum samples in leaf (default: 20)
#' @param subsample Row subsampling fraction for trees (default: 1.0, no subsampling)
#' @param seed Random seed (default: NULL)
#' @param verbose Print progress (default: FALSE)
#'
#' @return An lgb_plus_A model object (list)
#'
#' @examples
#' \dontrun{
#' X <- matrix(rnorm(1000 * 4), ncol = 4)
#' y <- 0.5 * X[,1] + 0.3 * tanh(X[,2]) + rnorm(1000) * 0.1
#' model <- lgb_plus_A(X, y, subsample = 0.75)
#' pred <- predict(model, X)
#' }
#'
#' @export
lgb_plus_A <- function(
    X, y,
    n_cycles = 25,
    trees_per_cycle = 10,
    lr_tree = 0.02,
    lr_linear = 0.1,
    num_leaves = 15,
    min_data = 20,
    subsample = 1.0,
    seed = NULL,
    verbose = FALSE
) {
  .check_lightgbm()
  
  # Convert inputs
  X <- as.matrix(X)
  y <- as.numeric(y)
  
  n_samples <- nrow(X)
  n_features <- ncol(X)
  
  feature_names <- colnames(X)
  if (is.null(feature_names)) {
    feature_names <- paste0("V", seq_len(n_features))
  }
  
  # Initialize
  init_pred <- mean(y)
  pred <- rep(init_pred, n_samples)
  
  trees <- list()
  linear_steps <- list()
  linear_feature_counts <- rep(0, n_features)
  names(linear_feature_counts) <- feature_names
  
  # LightGBM parameters
  lgb_params <- list(
    objective = "regression",
    num_leaves = num_leaves,
    min_data_in_leaf = min_data,
    learning_rate = 1.0,  # We apply lr_tree manually
    verbosity = -1,
    force_col_wise = TRUE
  )

  # Add subsampling if < 1.0
  if (subsample < 1.0) {
    lgb_params$bagging_fraction <- subsample
    lgb_params$bagging_freq <- 1  # Subsample every iteration
  }

  if (!is.null(seed)) {
    lgb_params$seed <- seed
    lgb_params$bagging_seed <- seed  # For reproducible subsampling
  }
  
  # Main loop
  for (cycle in seq_len(n_cycles)) {
    # === Tree step ===
    resid <- y - pred
    
    dtrain <- lightgbm::lgb.Dataset(
      data = X,
      label = resid,
      free_raw_data = FALSE
    )
    
    booster <- lightgbm::lgb.train(
      params = lgb_params,
      data = dtrain,
      nrounds = trees_per_cycle,
      verbose = -1
    )
    
    tree_pred <- predict(booster, X)
    pred <- pred + lr_tree * tree_pred
    trees[[cycle]] <- booster
    
    # === Linear step ===
    resid <- y - pred
    
    # Find feature most correlated with residual
    correlations <- sapply(seq_len(n_features), function(j) {
      cor(X[, j], resid, use = "complete.obs")
    })
    correlations[is.na(correlations)] <- 0
    
    best_j <- which.max(abs(correlations))
    
    # Fit simple linear regression
    x_j <- X[, best_j]
    x_mean <- mean(x_j)
    x_var <- var(x_j)
    
    if (x_var > 1e-10) {
      coef <- cov(x_j, resid) / x_var
      intercept <- mean(resid) - coef * x_mean
    } else {
      coef <- 0
      intercept <- mean(resid)
    }
    
    linear_pred <- coef * x_j + intercept
    pred <- pred + lr_linear * linear_pred
    
    linear_steps[[cycle]] <- list(
      feature = best_j,
      coef = coef,
      intercept = intercept
    )
    linear_feature_counts[best_j] <- linear_feature_counts[best_j] + 1
    
    if (verbose && cycle %% 5 == 0) {
      train_rmse <- sqrt(mean((y - pred)^2))
      cat(sprintf("Cycle %d/%d: RMSE=%.4f, linear feature=%s\n",
                  cycle, n_cycles, train_rmse, feature_names[best_j]))
    }
  }
  
  # Return model object
  model <- list(
    init = init_pred,
    trees = trees,
    linear_steps = linear_steps,
    n_cycles = n_cycles,
    trees_per_cycle = trees_per_cycle,
    lr_tree = lr_tree,
    lr_linear = lr_linear,
    subsample = subsample,
    n_features = n_features,
    feature_names = feature_names,
    linear_feature_counts = linear_feature_counts
  )
  
  class(model) <- "lgb_plus_A"
  return(model)
}


#' Predict method for lgb_plus_A
#'
#' @param object An lgb_plus_A model
#' @param newdata Matrix or data.frame of features
#' @param ... Additional arguments (ignored)
#'
#' @return Numeric vector of predictions
#'
#' @export
predict.lgb_plus_A <- function(object, newdata, ...) {
  .check_lightgbm()
  
  X <- as.matrix(newdata)
  n_samples <- nrow(X)
  
  pred <- rep(object$init, n_samples)
  
  for (cycle in seq_along(object$trees)) {
    booster <- object$trees[[cycle]]
    lin <- object$linear_steps[[cycle]]
    
    # Tree prediction
    pred <- pred + object$lr_tree * predict(booster, X)
    
    # Linear prediction
    j <- lin$feature
    pred <- pred + object$lr_linear * (lin$coef * X[, j] + lin$intercept)
  }
  
  return(pred)
}


#' Print method for lgb_plus_A
#'
#' @param x An lgb_plus_A model
#' @param ... Additional arguments (ignored)
#'
#' @export
print.lgb_plus_A <- function(x, ...) {
  cat("LGB^A+ Model\n")
  cat("=======================\n")
  cat(sprintf("Cycles: %d\n", x$n_cycles))
  cat(sprintf("Trees per cycle: %d\n", x$trees_per_cycle))
  cat(sprintf("Total trees: %d\n", x$n_cycles * x$trees_per_cycle))
  cat(sprintf("Learning rates: tree=%.3f, linear=%.3f\n", x$lr_tree, x$lr_linear))
  cat(sprintf("Features: %d\n", x$n_features))
  cat("\nLinear feature selection frequency:\n")
  
  sorted_counts <- sort(x$linear_feature_counts, decreasing = TRUE)
  for (i in seq_along(sorted_counts)) {
    if (sorted_counts[i] > 0) {
      cat(sprintf("  %s: %d times\n", names(sorted_counts)[i], sorted_counts[i]))
    }
  }
  
  invisible(x)
}


#' Summary method for lgb_plus_A
#'
#' @param object An lgb_plus_A model
#' @param ... Additional arguments (ignored)
#'
#' @return Summary statistics (invisibly)
#'
#' @export
summary.lgb_plus_A <- function(object, ...) {
  print(object)
  
  cat("\nLinear coefficients (cumulative, scaled by lr_linear):\n")
  
  # Aggregate linear contributions by feature
  linear_contrib <- rep(0, object$n_features)
  names(linear_contrib) <- object$feature_names
  
  for (lin in object$linear_steps) {
    linear_contrib[lin$feature] <- linear_contrib[lin$feature] + 
      object$lr_linear * lin$coef
  }
  
  sorted_contrib <- sort(abs(linear_contrib), decreasing = TRUE)
  for (i in seq_along(sorted_contrib)) {
    fname <- names(sorted_contrib)[i]
    val <- linear_contrib[fname]
    if (abs(val) > 1e-6) {
      cat(sprintf("  %s: %.4f\n", fname, val))
    }
  }
  
  invisible(list(
    linear_feature_counts = object$linear_feature_counts,
    linear_contributions = linear_contrib
  ))
}


#' Cross-validation for lgb_plus_A
#'
#' @param X Matrix or data.frame of features
#' @param y Numeric vector of target values
#' @param n_folds Number of CV folds (default: 5)
#' @param ... Additional arguments passed to lgb_plus_A
#'
#' @return List with CV results
#'
#' @export
cv_lgb_plus_A <- function(X, y, n_folds = 5, ...) {
  X <- as.matrix(X)
  y <- as.numeric(y)
  n <- nrow(X)
  
  # Create folds
  fold_ids <- sample(rep(1:n_folds, length.out = n))
  
  rmse_folds <- numeric(n_folds)
  
  for (fold in seq_len(n_folds)) {
    test_idx <- which(fold_ids == fold)
    train_idx <- which(fold_ids != fold)
    
    X_train <- X[train_idx, , drop = FALSE]
    y_train <- y[train_idx]
    X_test <- X[test_idx, , drop = FALSE]
    y_test <- y[test_idx]
    
    model <- lgb_plus_A(X_train, y_train, ...)
    pred <- predict(model, X_test)
    
    rmse_folds[fold] <- sqrt(mean((y_test - pred)^2))
  }
  
  list(
    rmse_mean = mean(rmse_folds),
    rmse_sd = sd(rmse_folds),
    rmse_folds = rmse_folds
  )
}


# ============================================================
# Demo / Test
# ============================================================

#' Fit an Ensembled LGB^A+ model
#'
#' Runs multiple lgb_plus_A models with tree subsampling and averages predictions.
#' The ensemble helps reduce variance while the subsampling prevents overfitting.
#'
#' @param X Matrix or data.frame of features (n_samples x n_features)
#' @param y Numeric vector of target values
#' @param n_runs Number of ensemble members (default: 10)
#' @param subsample_rate Subsampling rate for trees (default: 0.7)
#' @param n_cycles Number of alternating cycles (default: 25)
#' @param trees_per_cycle Trees to fit per cycle (default: 10)
#' @param lr_tree Learning rate for trees (default: 0.02)
#' @param lr_linear Learning rate for linear updates (default: 0.1)
#' @param num_leaves Max leaves per tree (default: 15)
#' @param min_data Minimum samples in leaf (default: 20)
#' @param seed Random seed (default: NULL)
#' @param verbose Print progress (default: FALSE)
#'
#' @return An lgb_plus_A_ensemble model object (list)
#'
#' @examples
#' \dontrun{
#' X <- matrix(rnorm(1000 * 4), ncol = 4)
#' y <- 0.5 * X[,1] + 0.3 * tanh(X[,2]) + rnorm(1000) * 0.1
#' model <- lgb_plus_A_ensemble(X, y, n_runs = 10, subsample_rate = 0.7)
#' pred <- predict(model, X)
#' }
#'
#' @export
lgb_plus_A_ensemble <- function(
    X, y,
    n_runs = 10,
    subsample_rate = 0.7,
    n_cycles = 25,
    trees_per_cycle = 10,
    lr_tree = 0.02,
    lr_linear = 0.1,
    num_leaves = 15,
    min_data = 20,
    seed = NULL,
    verbose = FALSE
) {
  .check_lightgbm()

  # Convert inputs
  X <- as.matrix(X)
  y <- as.numeric(y)

  n_samples <- nrow(X)
  n_features <- ncol(X)

  feature_names <- colnames(X)
  if (is.null(feature_names)) {
    feature_names <- paste0("V", seq_len(n_features))
  }

  # Storage for ensemble models
  ensemble_models <- list()

  # Aggregate linear feature counts across ensemble
  linear_feature_counts_total <- rep(0, n_features)
  names(linear_feature_counts_total) <- feature_names

  # LightGBM parameters with subsampling for trees
  lgb_params <- list(
    objective = "regression",
    num_leaves = num_leaves,
    min_data_in_leaf = min_data,
    learning_rate = 1.0,  # We apply lr_tree manually
    verbosity = -1,
    force_col_wise = TRUE,
    bagging_fraction = subsample_rate,  # Subsample rate for trees
    bagging_freq = 1                     # Apply bagging every iteration
  )

  # Main ensemble loop
  for (run in seq_len(n_runs)) {
    if (!is.null(seed)) {
      run_seed <- seed + run - 1
    } else {
      run_seed <- NULL
    }

    if (verbose) {
      cat(sprintf("Fitting ensemble member %d/%d\n", run, n_runs))
    }

    # Initialize for this run
    init_pred <- mean(y)
    pred <- rep(init_pred, n_samples)

    trees <- list()
    linear_steps <- list()
    linear_feature_counts <- rep(0, n_features)
    names(linear_feature_counts) <- feature_names

    # Set seed for this run
    if (!is.null(run_seed)) {
      lgb_params$seed <- run_seed
      lgb_params$bagging_seed <- run_seed
    }

    # Cycle loop
    for (cycle in seq_len(n_cycles)) {
      # === Tree step ===
      resid <- y - pred

      dtrain <- lightgbm::lgb.Dataset(
        data = X,
        label = resid,
        free_raw_data = FALSE
      )

      booster <- lightgbm::lgb.train(
        params = lgb_params,
        data = dtrain,
        nrounds = trees_per_cycle,
        verbose = -1
      )

      tree_pred <- predict(booster, X)
      pred <- pred + lr_tree * tree_pred
      trees[[cycle]] <- booster

      # === Linear step ===
      resid <- y - pred

      # Find feature most correlated with residual
      correlations <- sapply(seq_len(n_features), function(j) {
        cor(X[, j], resid, use = "complete.obs")
      })
      correlations[is.na(correlations)] <- 0

      best_j <- which.max(abs(correlations))

      # Fit simple linear regression
      x_j <- X[, best_j]
      x_mean <- mean(x_j)
      x_var <- var(x_j)

      if (x_var > 1e-10) {
        coef <- cov(x_j, resid) / x_var
        intercept <- mean(resid) - coef * x_mean
      } else {
        coef <- 0
        intercept <- mean(resid)
      }

      linear_pred <- coef * x_j + intercept
      pred <- pred + lr_linear * linear_pred

      linear_steps[[cycle]] <- list(
        feature = best_j,
        coef = coef,
        intercept = intercept
      )
      linear_feature_counts[best_j] <- linear_feature_counts[best_j] + 1
    }

    # Store this run's model
    ensemble_models[[run]] <- list(
      init = init_pred,
      trees = trees,
      linear_steps = linear_steps,
      linear_feature_counts = linear_feature_counts
    )

    # Accumulate feature counts
    linear_feature_counts_total <- linear_feature_counts_total + linear_feature_counts
  }

  # Return ensemble model object
  model <- list(
    ensemble_models = ensemble_models,
    n_runs = n_runs,
    subsample_rate = subsample_rate,
    n_cycles = n_cycles,
    trees_per_cycle = trees_per_cycle,
    lr_tree = lr_tree,
    lr_linear = lr_linear,
    n_features = n_features,
    feature_names = feature_names,
    linear_feature_counts = linear_feature_counts_total / n_runs  # Average counts
  )

  class(model) <- "lgb_plus_A_ensemble"
  return(model)
}


#' Predict method for lgb_plus_A_ensemble
#'
#' @param object An lgb_plus_A_ensemble model
#' @param newdata Matrix or data.frame of features
#' @param ... Additional arguments (ignored)
#'
#' @return Numeric vector of predictions (averaged across ensemble)
#'
#' @export
predict.lgb_plus_A_ensemble <- function(object, newdata, ...) {
  .check_lightgbm()

  X <- as.matrix(newdata)
  n_samples <- nrow(X)

  # Accumulate predictions from all ensemble members
  pred_sum <- rep(0, n_samples)

  for (run in seq_along(object$ensemble_models)) {
    member <- object$ensemble_models[[run]]
    pred <- rep(member$init, n_samples)

    for (cycle in seq_along(member$trees)) {
      booster <- member$trees[[cycle]]
      lin <- member$linear_steps[[cycle]]

      # Tree prediction
      pred <- pred + object$lr_tree * predict(booster, X)

      # Linear prediction
      j <- lin$feature
      pred <- pred + object$lr_linear * (lin$coef * X[, j] + lin$intercept)
    }

    pred_sum <- pred_sum + pred
  }

  # Average predictions
  return(pred_sum / object$n_runs)
}


#' Print method for lgb_plus_A_ensemble
#'
#' @param x An lgb_plus_A_ensemble model
#' @param ... Additional arguments (ignored)
#'
#' @export
print.lgb_plus_A_ensemble <- function(x, ...) {
  cat("LGB^A+ Ensemble Model\n")
  cat("=================================\n")
  cat(sprintf("Ensemble members: %d\n", x$n_runs))
  cat(sprintf("Tree subsample rate: %.2f\n", x$subsample_rate))
  cat(sprintf("Cycles per member: %d\n", x$n_cycles))
  cat(sprintf("Trees per cycle: %d\n", x$trees_per_cycle))
  cat(sprintf("Total trees per member: %d\n", x$n_cycles * x$trees_per_cycle))
  cat(sprintf("Learning rates: tree=%.3f, linear=%.3f\n", x$lr_tree, x$lr_linear))
  cat(sprintf("Features: %d\n", x$n_features))
  cat("\nAvg linear feature selection frequency:\n")

  sorted_counts <- sort(x$linear_feature_counts, decreasing = TRUE)
  for (i in seq_along(sorted_counts)) {
    if (sorted_counts[i] > 0) {
      cat(sprintf("  %s: %.1f times\n", names(sorted_counts)[i], sorted_counts[i]))
    }
  }

  invisible(x)
}


if (FALSE) {  # Set to TRUE to run demo

  set.seed(42)
  
  # Generate hybrid DGP data
  n <- 2000
  X <- matrix(rnorm(n * 4), ncol = 4)
  colnames(X) <- c("x1", "x2", "x3", "x4")
  
  # y = linear + nonlinear + noise
  y <- 0.5 * X[, 1] +                      # linear
    0.3 * tanh(2 * X[, 2]) +               # nonlinear
    0.2 * X[, 1] * (X[, 2] > 0) +          # interaction
    rnorm(n) * 0.3                         # noise
  
  # Train/test split
  train_idx <- 1:1500
  test_idx <- 1501:2000
  
  X_train <- X[train_idx, ]
  y_train <- y[train_idx]
  X_test <- X[test_idx, ]
  y_test <- y[test_idx]
  
  # Fit model
  model <- lgb_plus_A(X_train, y_train, verbose = TRUE)
  
  # Predict
  pred_train <- predict(model, X_train)
  pred_test <- predict(model, X_test)
  
  # Evaluate
  cat("\nResults:\n")
  cat(sprintf("Train RMSE: %.4f\n", sqrt(mean((y_train - pred_train)^2))))
  cat(sprintf("Test RMSE:  %.4f\n", sqrt(mean((y_test - pred_test)^2))))
  
  # Summary
  cat("\n")
  print(model)
  
  # Cross-validation
  cat("\nCross-validation:\n")
  cv_result <- cv_lgb_plus_A(X, y, n_folds = 5)
  cat(sprintf("CV RMSE: %.4f (+/- %.4f)\n", cv_result$rmse_mean, cv_result$rmse_sd))
}

File: python/lgb_plus_A.py
"""
LGB^A+: Hybrid Tree + Linear Boosting

A gradient boosting variant that alternates between:
1. Fitting a block of trees on residuals
2. Fitting a single linear term (greedy feature selection) on residuals

This captures both nonlinear patterns (via trees) and linear structure 
(via repeated linear updates), making it robust to unknown DGPs.

Author: Philippe Goulet Coulombe (UQAM)
License: MIT
"""

import numpy as np
from typing import Optional, List, Tuple
import warnings


class LGBPlusA:
    """
    LGB^A+ regressor.
    
    Alternates between gradient boosting trees and greedy linear updates.
    Effective when the true DGP has both linear and nonlinear components.
    
    Parameters
    ----------
    n_cycles : int, default=25
        Number of alternating cycles. Each cycle fits one block of trees
        followed by one linear update.
        
    trees_per_cycle : int, default=10
        Number of trees to fit per cycle. Total trees = n_cycles × trees_per_cycle.
        
    lr_tree : float, default=0.02
        Learning rate (shrinkage) for tree predictions.
        
    lr_linear : float, default=0.1
        Learning rate for linear updates. Higher values let linear
        component build up faster.
        
    num_leaves : int, default=15
        Maximum number of leaves per tree. Lower = more regularization.
        
    min_child_samples : int, default=20
        Minimum samples required in a leaf.
        
    random_state : int or None, default=None
        Random seed for reproducibility.
        
    verbose : bool, default=False
        If True, print progress during fitting.
    
    Attributes
    ----------
    init_ : float
        Initial prediction (mean of training target).
        
    trees_ : list
        Fitted tree boosters, one per cycle.
        
    linear_steps_ : list of tuples
        Each tuple is (feature_index, coefficient, intercept) for the
        linear update in that cycle.
        
    feature_importances_ : ndarray of shape (n_features,)
        Feature importances from trees (sum of gains).
        
    linear_feature_counts_ : ndarray of shape (n_features,)
        How many times each feature was selected for linear updates.
    
    Examples
    --------
    >>> from lgb_plus_A import LGBPlusA
    >>> import numpy as np
    >>> X = np.random.randn(1000, 5)
    >>> y = 0.5 * X[:, 0] + 0.3 * np.tanh(X[:, 1]) + np.random.randn(1000) * 0.1
    >>> model = LGBPlusA(n_cycles=25, trees_per_cycle=10)
    >>> model.fit(X, y)
    >>> predictions = model.predict(X)
    """
    
    def __init__(
        self,
        n_cycles: int = 25,
        trees_per_cycle: int = 10,
        lr_tree: float = 0.02,
        lr_linear: float = 0.1,
        num_leaves: int = 15,
        min_child_samples: int = 20,
        random_state: Optional[int] = None,
        verbose: bool = False
    ):
        self.n_cycles = n_cycles
        self.trees_per_cycle = trees_per_cycle
        self.lr_tree = lr_tree
        self.lr_linear = lr_linear
        self.num_leaves = num_leaves
        self.min_child_samples = min_child_samples
        self.random_state = random_state
        self.verbose = verbose
        
    def fit(self, X, y, feature_names: Optional[List[str]] = None):
        """
        Fit the alternating boost model.
        
        Parameters
        ----------
        X : array-like of shape (n_samples, n_features)
            Training features.
            
        y : array-like of shape (n_samples,)
            Training target.
            
        feature_names : list of str, optional
            Names for features (used in verbose output).
            
        Returns
        -------
        self : LGBPlusA
            Fitted model.
        """
        try:
            import lightgbm as lgb
        except ImportError:
            raise ImportError("LightGBM is required. Install with: pip install lightgbm")
        
        X = np.asarray(X, dtype=np.float64)
        y = np.asarray(y, dtype=np.float64).ravel()
        
        n_samples, n_features = X.shape
        
        if feature_names is None:
            feature_names = [f"x{i}" for i in range(n_features)]
        
        # Initialize
        self.init_ = float(y.mean())
        pred = np.full(n_samples, self.init_)
        
        self.trees_ = []
        self.linear_steps_ = []
        self.feature_importances_ = np.zeros(n_features)
        self.linear_feature_counts_ = np.zeros(n_features, dtype=int)
        
        # LightGBM parameters
        lgb_params = dict(
            objective="regression",
            num_leaves=self.num_leaves,
            min_child_samples=self.min_child_samples,
            verbosity=-1,
            force_col_wise=True,
        )
        if self.random_state is not None:
            lgb_params["seed"] = self.random_state
        
        for cycle in range(self.n_cycles):
            # === Tree step ===
            resid = y - pred
            dtrain = lgb.Dataset(X, label=resid, free_raw_data=False)
            
            with warnings.catch_warnings():
                warnings.simplefilter("ignore")
                booster = lgb.train(lgb_params, dtrain, self.trees_per_cycle)
            
            tree_pred = booster.predict(X)
            pred = pred + self.lr_tree * tree_pred
            self.trees_.append(booster)
            
            # Accumulate feature importance
            importance = booster.feature_importance(importance_type='gain')
            if len(importance) == n_features:
                self.feature_importances_ += importance
            
            # === Linear step ===
            resid = y - pred
            
            # Find feature most correlated with residual
            best_j, best_corr = 0, 0.0
            for j in range(n_features):
                corr = np.corrcoef(X[:, j], resid)[0, 1]
                if np.isnan(corr):
                    corr = 0.0
                if abs(corr) > abs(best_corr):
                    best_corr = corr
                    best_j = j
            
            # Fit simple linear regression on that feature
            x_j = X[:, best_j]
            x_mean, x_var = x_j.mean(), x_j.var()
            if x_var > 1e-10:
                coef = np.cov(x_j, resid)[0, 1] / x_var
                intercept = resid.mean() - coef * x_mean
            else:
                coef, intercept = 0.0, resid.mean()
            
            linear_pred = coef * x_j + intercept
            pred = pred + self.lr_linear * linear_pred
            
            self.linear_steps_.append((best_j, coef, intercept))
            self.linear_feature_counts_[best_j] += 1
            
            if self.verbose and (cycle + 1) % 5 == 0:
                train_rmse = np.sqrt(np.mean((y - pred) ** 2))
                print(f"Cycle {cycle+1}/{self.n_cycles}: "
                      f"RMSE={train_rmse:.4f}, "
                      f"linear feature={feature_names[best_j]}")
        
        return self
    
    def predict(self, X) -> np.ndarray:
        """
        Predict using the fitted model.
        
        Parameters
        ----------
        X : array-like of shape (n_samples, n_features)
            Features to predict.
            
        Returns
        -------
        y_pred : ndarray of shape (n_samples,)
            Predicted values.
        """
        X = np.asarray(X, dtype=np.float64)
        pred = np.full(X.shape[0], self.init_)
        
        for booster, (j, coef, intercept) in zip(self.trees_, self.linear_steps_):
            pred = pred + self.lr_tree * booster.predict(X)
            pred = pred + self.lr_linear * (coef * X[:, j] + intercept)
        
        return pred
    
    def get_total_trees(self) -> int:
        """Return total number of trees in the ensemble."""
        return self.n_cycles * self.trees_per_cycle
    
    def summary(self) -> str:
        """Return a summary of the fitted model."""
        if not hasattr(self, 'trees_'):
            return "Model not fitted yet."
        
        lines = [
            "LGBPlusA Summary",
            "=" * 40,
            f"Total cycles: {self.n_cycles}",
            f"Trees per cycle: {self.trees_per_cycle}",
            f"Total trees: {self.get_total_trees()}",
            f"Learning rates: tree={self.lr_tree}, linear={self.lr_linear}",
            "",
            "Linear feature selection frequency:",
        ]
        
        # Sort by count
        sorted_idx = np.argsort(self.linear_feature_counts_)[::-1]
        for idx in sorted_idx:
            if self.linear_feature_counts_[idx] > 0:
                lines.append(f"  Feature {idx}: {self.linear_feature_counts_[idx]} times")
        
        return "\n".join(lines)


# Convenience function for quick fitting
def lgb_plus_A(
    X_train, y_train, X_test=None,
    n_cycles: int = 25,
    trees_per_cycle: int = 10,
    lr_tree: float = 0.02,
    lr_linear: float = 0.1,
    random_state: Optional[int] = None
) -> Tuple[LGBPlusA, np.ndarray, Optional[np.ndarray]]:
    """
    Convenience function to fit LGBPlusA and get predictions.
    
    Parameters
    ----------
    X_train : array-like
        Training features.
    y_train : array-like
        Training target.
    X_test : array-like, optional
        Test features. If provided, returns test predictions.
    n_cycles : int
        Number of alternating cycles.
    trees_per_cycle : int
        Trees per cycle.
    lr_tree : float
        Tree learning rate.
    lr_linear : float
        Linear learning rate.
    random_state : int, optional
        Random seed.
        
    Returns
    -------
    model : LGBPlusA
        Fitted model.
    train_pred : ndarray
        Training predictions.
    test_pred : ndarray or None
        Test predictions (if X_test provided).
    """
    model = LGBPlusA(
        n_cycles=n_cycles,
        trees_per_cycle=trees_per_cycle,
        lr_tree=lr_tree,
        lr_linear=lr_linear,
        random_state=random_state
    )
    model.fit(X_train, y_train)
    
    train_pred = model.predict(X_train)
    test_pred = model.predict(X_test) if X_test is not None else None
    
    return model, train_pred, test_pred


if __name__ == "__main__":
    # Quick demo
    print("LGBPlusA Demo")
    print("=" * 50)
    
    np.random.seed(42)
    
    # Generate hybrid DGP data
    n = 2000
    X = np.random.randn(n, 4)
    
    # y = linear + nonlinear + noise
    y = (0.5 * X[:, 0] +                    # linear
         0.3 * np.tanh(2 * X[:, 1]) +       # nonlinear
         0.2 * X[:, 0] * (X[:, 1] > 0) +    # interaction
         np.random.randn(n) * 0.3)          # noise
    
    # Train/test split
    X_train, X_test = X[:1500], X[1500:]
    y_train, y_test = y[:1500], y[1500:]
    
    # Fit model
    model = LGBPlusA(n_cycles=25, trees_per_cycle=10, verbose=True)
    model.fit(X_train, y_train, feature_names=['x0', 'x1', 'x2', 'x3'])
    
    # Evaluate
    from sklearn.metrics import root_mean_squared_error
    
    pred_train = model.predict(X_train)
    pred_test = model.predict(X_test)
    
    print(f"\nTrain RMSE: {root_mean_squared_error(y_train, pred_train):.4f}")
    print(f"Test RMSE:  {root_mean_squared_error(y_test, pred_test):.4f}")
    print(f"\n{model.summary()}")
